File: src/main/scala/ao/learn/mst/gen5/state/strategy/impl/AverageStrategy.scala
package ao.learn.mst.gen5.state.strategy.impl

import ao.learn.mst.gen5.state.MixedStrategy
import ao.learn.mst.gen5.state.strategy.StrategyAccumulator

/**
 *
 */
class AverageStrategy(strategyAccumulator: StrategyAccumulator)
    extends MixedStrategy
{
  override def probabilities(informationSetIndex: Long, actionCount: Int): IndexedSeq[Double] = {
    val probabilitySums: IndexedSeq[Double] =
      strategyAccumulator.cumulativeStrategy(informationSetIndex)

    if (probabilitySums.isEmpty)
    {
      // default; todo: can this be eliminated?
      IndexedSeq.fill(actionCount)(1.0 / actionCount)
    }
    else
    {
      val strategyTotal: Double =
        probabilitySums.sum

      val averageStrategy: IndexedSeq[Double] =
        probabilitySums
          .map(_ / strategyTotal)

      require(averageStrategy.length <= actionCount)

      if (averageStrategy.length == actionCount) {
        averageStrategy
      } else {
        averageStrategy.padTo(actionCount, 0.0)
      }
    }
  }
}

File: ai-cfr-java/src/main/java/ao/ai/cfr/state/AverageStrategy.java
package ao.ai.cfr.state;


import java.util.Arrays;

public class AverageStrategy implements MixedStrategy
{
    private final StrategyAccumulator strategyAccumulator;


    public AverageStrategy(StrategyAccumulator strategyAccumulator) {
        this.strategyAccumulator = strategyAccumulator;
    }


    @Override
    public double[] probabilities(int informationSetIndex, int actionCount) {
        assert actionCount > 0;

        double[] probabilitySums =
                strategyAccumulator.cumulativeStrategy(informationSetIndex);

        if (probabilitySums == null || probabilitySums.length == 0)
        {
            // default; todo: can this be eliminated?
            double[] uniform = new double[actionCount];
            Arrays.fill(uniform, 1.0 / actionCount);
            return uniform;
        }

        assert probabilitySums.length <= actionCount;

        double strategyTotal = 0;
        for (double probabilitySum : probabilitySums) {
            strategyTotal += probabilitySum;
        }

        for (int i = 0; i < probabilitySums.length; i++) {
            probabilitySums[i] /= strategyTotal;
        }

        if (probabilitySums.length == actionCount) {
            return probabilitySums;
        } else {
            return Arrays.copyOf(probabilitySums, actionCount);
        }
    }
}
